lib/boss/participant.rb

```diff
@@ -0,0 +1,357 @@
+  #
+  # By default, 'exchange' is set to [ 'direct', '' ] (the default exchange).
+  #
+  # Note: you cannot pass an instantiated Ruby-AMQP exchange here. Ruote
+  # cannot serialize it for remote workers, so the settings are passed
+  # in a flat form, easily JSONifiable.
+  #
+  # === 'field_prefix'
+  #
+  # Sometimes one wants to separate his AMQP participant settings from other
+  # workitem fields.
+  #
+  #   dashboard.register(
+  #     'amqp_participant',
+  #     BOSS::Participant,
+  #     :conf => 'fields', :field_prefix => 'amqp_')
+  #
+  # registers a participant that draws is configuration from workitem fields
+  # prefixed with 'amqp_'.
+  #
+  # Note that setting this option doesn't implicitely add 'fields' to the
+  # 'conf' option.
+  #
+  # === 'forget'
+  #
+  # When set to true forces the participant to reply to the engine immediately
+  # after the message got published, in a "fire and forget" fashion.
+  #
+  # === 'routing_key'
+  #
+  # Depending on the exchange used, this option lets you influence how the
+  # exchange routes the message towards queues.
+  #
+  # Consult your AMQP documentation for more information.
+  #
+  # === 'message'
+  #
+  # By default, the workitem is turned into a JSON string and transmitted in
+  # the AMQP message payload. If this 'message' option is set, its value is
+  # used as the payload.
+  #
+  # === 'persistent'
+  #
+  # If this option is set to something else than false or nil, messages
+  # messages published by this participant will be persistent (hopefully
+  # the queues they'll end up in will be persistent as well).
+  #
+  #
+  # == #encode_workitem
+  #
+  # The default way to encode a workitem before pushing it to the exchange
+  # is by turning it entirely into a JSON string.
+  #
+  # To alter this, one can subclass this participant and provide its own
+  # #encode_workitem(wi) method:
+  #
+  #   require 'yaml'
+  #
+  #   class MyAmqpParticipant < BOSS::Participant
+  #
+  #     def encode_workitem(workitem)
+  #       YAML.dump(workitem)
+  #     end
+  #   end
+  #
+  # or when one needs to filter some fields:
+  #
+  #   class MyAmqpParticipant < BOSS::Participant
+  #
+  #     def encode_workitem(workitem)
+  #       workitem.fields.delete_if { |k, v| k.match(/^private_/) }
+  #       super(workitem)
+  #     end
+  #   end
+  #
+  class Participant
+    include Ruote::LocalParticipant
+
+    # Initializing the participant, right before calling #on_workitem or
+    # another on_ method.
+    #
+    def initialize(options)
+
+      @options = options
+
+      @conf = (@options['conf'] || 'params, fields, options').split(/\s*,\s*/)
+      @conf = %w[ params fields options ] if @conf.include?('all')
+
+      @field_prefix = @options['field_prefix'] || ''
+    end
+
+    # Workitem consumption code.
+    #
+    def on_workitem
+
+      begin
+        instantiate_exchange.publish(
+                                     message,
```

Comment:
Inconsistent indentation compared to the one in on_cancel. And you know my feelings towards the indentation aligned with the opening :wink:

Reply:
tbh in ruby I don't know what style to use (and sorry, I forget your preferences ;) )
I press 'Tab' in emacs and let emacs choose how to indent any code I write.
I tend not to re-indent the whole file. I've made on_cancel match now...